js/style/style.js

```diff
@@ -638,7 +638,16 @@ Style.prototype = util.inherit(Evented, {
         }
 
         var sourceResults = [];
-        for (var id in this.sources) {
+        var sourcesToQuery = {};
+        if (params.layers) {
+            for (var i = 0; i < params.layers.length; i++) {
+                var sourceId = this._layers[params.layers[i]].source;
+                sourcesToQuery[sourceId] = true;
+            }
+        } else {
+            sourcesToQuery = this.sources;
+        }
+        for (var id in sourcesToQuery) {
```

Comment:
Given that this isn't hot code, do you think we could get by with a simpler implementation like this 👇 ?

```js
for (var id in this.sources) {
    if (!params.layers || params.layers.some(function(layerId) { return this._layers[layerId].source === id; }) continue;
    ...
```